thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/anomaly/detection/trigger/DataAvailabilityTaskScheduler.java

```diff
@@ -0,0 +1,222 @@
+  private int maxNotRunThresholdInMin; // default threshold if detection level threshold is not set
+  private Map<Long, Long> lastTaskCreateTimeMap;
+  private TaskManager taskDAO;
+  private DetectionConfigManager detectionConfigDAO;
+  private DatasetConfigManager datasetConfigDAO;
+  /**
+   * Construct an instance of {@link DataAvailabilityTaskScheduler}
+   * @param delayInSec delay after each run to avoid polling the database too often
+   * @param maxNotRunThresholdInMin global threshold for fallback if detection level one is not set
+   */
+  public DataAvailabilityTaskScheduler(long delayInSec, int maxNotRunThresholdInMin) {
+    this.delayInSec = delayInSec;
+    this.maxNotRunThresholdInMin = maxNotRunThresholdInMin;
+    this.lastTaskCreateTimeMap = new HashMap<>();
+    this.executorService = Executors.newSingleThreadScheduledExecutor();
+    this.taskDAO = DAORegistry.getInstance().getTaskDAO();
+    this.detectionConfigDAO = DAORegistry.getInstance().getDetectionConfigManager();
+    this.datasetConfigDAO = DAORegistry.getInstance().getDatasetConfigDAO();
+  }
+
+  @Override
+  public void run() {
+    Map<DetectionConfigDTO, Set<String>> detection2DatasetMap = new HashMap<>();
+    Map<String, Long> dataset2RefreshTimeMap = new HashMap<>();
+    populateDetectionMapAndDataset2RefreshTimeMap(detection2DatasetMap, dataset2RefreshTimeMap);
+    Map<Long, TaskDTO> runningDetection = retrieveRunningDetection();
+    int taskCount = 0;
+    for (DetectionConfigDTO detectionConfig : detection2DatasetMap.keySet()) {
+      try {
+        long detectionConfigId = detectionConfig.getId();
+        if (!runningDetection.containsKey(detectionConfigId)) {
+          if (isAllDatasetUpdated(detectionConfig, detection2DatasetMap.get(detectionConfig), dataset2RefreshTimeMap)) {
+            //TODO: additional check is required if detection is based on aggregated value across multiple data points
+            createDetectionTask(detectionConfig);
+            taskCount++;
+          } else if (isTimeForFallback(detectionConfig)) {
+            LOG.info("Scheduling a task for detection {} due to the fallback mechanism", detectionConfigId);
+            createDetectionTask(detectionConfig);
+            taskCount++;
+          }
+        } else {
+          LOG.info("Skipping creating detection task for detection {} because task {} is not finished.",
+              detectionConfigId, runningDetection.get(detectionConfigId));
+        }
+      } catch (Exception e) {
+        LOG.error("Error in scheduling a detection...", e);
+      }
+    }
+    ThirdeyeMetricsUtil.eventDrivenScheduledTaskCounter.inc(taskCount);
+    LOG.info("Schedule {} tasks in this run...", taskCount);
+
+  }
+
+  public void start() {
+    executorService.scheduleWithFixedDelay(this, 0, delayInSec, TimeUnit.SECONDS);
+  }
+
+  public void close() {
+    executorService.shutdownNow();
+  }
+
+  private void populateDetectionMapAndDataset2RefreshTimeMap(
```

Comment:
The name is too long...
Can we split it into two functions?
1. initDetectionDatasetMap()
2. initDatasetRefreshTimeMap()

Reply:
We can split them but it only moves last six lines to another methods. Since the initialization of  `dataset2RefreshTimeMap` depends on the initialization of `detection2DatasetMap`, should we put them into the same method?